File: ml/tests/matlab/svm/test3_kernel.m
% Mihail Dunaev
% May 2017
%
% Simple test that uses kernel SVM to separate between 2D data
% points (non-linear decision boundary). Data in 'logistic_circle.in'.

function test3_kernel()
    addpath('core/');
    [xs, ys] = read_data('../../input/logistic_circle.in');

    % random initial weights
    N = size(xs, 1);
    eps_theta = 0.1;
    w0 = rand(N+1, 1) * (2 * eps_theta) - eps_theta;
    
    syms t;
    C = 100;
    fun = @(t)(cost_with_grad_kernel(t, xs, ys, C));
    w = train(w0, fun);

    % predict some values
    xpred = [3.5 4; 4 4.1; 2 6];
    ypred = predict_kernel(xpred, w, xs);
    
    %figure;
    %hold on;
    %plot_kernels(xs);
    plot_all(xs, ys, xpred, ypred);
end

% plot similarity around each point
function plot_kernels(xs)
    N = size(xs, 1);
    delta = 0.1;
    for k = 1:N
        xmin = xs(k,1) - 1;
        xmax = xs(k,1) + 1;
        ymin = xs(k,2) - 1;
        ymax = xs(k,2) + 1;
        [wxs,wys] = meshgrid(xmin:delta:xmax, ymin:delta:ymax);
        sz = size(wxs);
        total_sz = sz(1)*sz(2);
        cwxs = reshape(wxs, total_sz, 1);
        cwys = reshape(wys, total_sz, 1);
        cwzs = zeros(total_sz, 1);
        for i = 1:total_sz
            cwzs(i,1) = kernel(xs(k,:), [cwxs(i) cwys(i)], 1);
        end
        wzs = reshape(cwzs, sz);
        surf(wxs, wys, wzs, 'EdgeColor', 'none');
    end
    alpha 0.1;
end

function plot_all(xs, ys, xpred, ypred)
    xs0 = xs(ys == 0,:);
    xs1 = xs(ys == 1,:);

    % plot initial data
    figure;
    hold on;
    
    %plot_similarity_test3();
    
    plot(xs0(:,1), xs0(:,2), 'bo');
    plot(xs1(:,1), xs1(:,2), 'rx');

    % set range for x/y axis
    xmin = min(xs(:,1));
    xmax = max(xs(:,1));
    ymin = min(xs(:,2));
    ymax = max(xs(:,2));
    xlim([xmin-1 xmax+1]);
    ylim([ymin-1 ymax+1]);
    
    % plot predicted points
    for i = 1:size(xpred, 1)
        if ypred(i) == 1
            plot(xpred(i,1), xpred(i,2), 'r+');
        elseif ypred(i) == 0
            plot(xpred(i,1), xpred(i,2), 'b+');
        else
            plot(xpred(i,1), xpred(i,2), 'k+');
        end
    end
end

File: ml/tests/python/scikit_learn/svm/test3_kernel.py
"""
Mihail Dunaev
May 2017

Simple test that uses kernel SVM to separate between 2D data
  points (non-linear decision boundary). Data in 'logistic_circle.in'.

For linear SVM I used scikit-learn:
  sklearn.svm.SVC
"""

import numpy as np
import matplotlib.pyplot as plt
from sklearn.svm import SVC
from matplotlib import colors
from matplotlib import cm

from common.aux import read_data

def main():
    """ wrapper for test3_kernel """
    test3_kernel(show_colormesh=True, nice_plotting=True)

def test3_kernel(show_colormesh=True, nice_plotting=True):
    """
    nice_plotting:

      computes the probabilities for SVM as well,
      which makes the training process slower
      (but the plot looks prettier).

      only works with show_colormesh = True.
    """
    (xs, ys) = read_data("../../../input/logistic_circle.in")
    if nice_plotting:
        svc = SVC(C=1e5, kernel="rbf", gamma='auto', probability=True)
    else:
        svc = SVC(C=1e5, kernel="rbf", gamma='auto')
    svc.fit(xs, ys)

    xpred = np.array([3.5, 4, 4, 4.1, 2, 6]).reshape(3, 2)
    ypred = svc.predict(xpred)
    print(ypred.tolist())

    plot_all(xs, ys, svc, xpred, ypred, show_colormesh, nice_plotting)

def plot_all(xs, ys, svc, xpred, ypred,
             show_colormesh=True, nice_plotting=True):

    # switch backend - TODO fix this
    plt.switch_backend("TKagg")

    N = len(xs)
    xs0 = np.array([xs[i] for i in xrange(N) if ys[i] == 0])
    xs1 = np.array([xs[i] for i in xrange(N) if ys[i] == 1])

    # plot initial data
    plt.plot(xs0[:,0], xs0[:,1], "bo")
    plt.plot(xs1[:,0], xs1[:,1], "rx")

    # set range for x/y axis
    xmin = xs[:,0].min()
    xmax = xs[:,0].max()
    ymin = xs[:,1].min()
    ymax = xs[:,1].max()
    plt.xlim([xmin - 1, xmax + 1])
    plt.ylim([ymin - 1, ymax + 1])

    # plot regions
    if show_colormesh:
        num_pts = 100
        my_cmap = colors.ListedColormap(["b", "r"])
        (wxs, wys) = np.meshgrid(np.linspace(xmin - 1, xmax + 1, num_pts),
                                 np.linspace(ymin - 1, ymax + 1, num_pts))
        if nice_plotting:
            wzs = svc.predict_proba(np.concatenate((wxs.reshape(wxs.size, 1),
                                    wys.reshape(wys.size, 1)), axis=1))[:,1]
            plt.imshow(wzs.reshape(wxs.shape), interpolation='bilinear',
                   origin='lower', cmap=my_cmap,
                   extent=(xmin - 1, xmax + 1, ymin - 1, ymax + 1),
                   alpha=0.5)
        else:
            wzs = svc.predict(np.concatenate((wxs.reshape(wxs.size, 1),
                                         wys.reshape(wys.size, 1)), axis=1))
            plt.pcolormesh(wxs, wys, wzs.reshape(wxs.shape), cmap=my_cmap,
                           alpha=0.5)

    # plot predicted points
    npred = xpred.shape[0]
    for i in xrange(npred):
        if ypred[i] == 0:
            plt.plot(xpred[i][0], xpred[i][1], "bs")
        elif ypred[i] == 1:
            plt.plot(xpred[i][0], xpred[i][1], "rs")
        else:
            plt.plot(xpred[i][0], xpred[i][1], "ks")
    plt.show()

main()
